src/pages/DoughStaking.svelte

```diff
@@ -0,0 +1,364 @@
+    }
+    
+  }
+
+  function checkApproval(allowance) {
+    if(stake.amount && allowance) {
+      stakeError = false;
+      let stakeAmount = BigNumber(stake.amount.toString()).multipliedBy(10 ** 18);
+
+      if(allowance.isGreaterThanOrEqualTo(stakeAmount)) {
+        needAllowance = false;
+      } else {
+        needAllowance = true;
+      }
+    } else {
+      stakeError = true;
+    }
+  }
+
+  async function approveToken() {
+    if (!$eth.address || !$eth.signer) {
+      displayNotification({ message: $_('piedao.please.connect.wallet'), type: 'hint' });
+      connectWeb3();
+      return;
+    }
+
+    try {
+      // resetting the approve to zero, before initiating a new approval...
+      if(!data.accountDepositTokenAllowance.isEqualTo(0)) {
+        await approve(smartcontracts.dough, smartcontracts.doughStaking, 0);
+      }
+
+      let stakeAmount = BigNumber(stake.amount.toString()).multipliedBy(10 ** 18);     
+      await approve(smartcontracts.dough, smartcontracts.doughStaking, stakeAmount.toString(), {gasLimit: 100000});
+      needAllowance = false; 
+      
+    } catch(error) {
+      stakeError = true;
+      needAllowance = true;
+    }
+  }
+
+  async function stakeDOUGH() {
+    if(!sharesTimeLock) return;
+
+    try {
+      let response = await sharesTimeLock.depositByMonths(
+        parseEther(stake.amount.toString()), 
+        stake.duration, 
+        stake.receiver);
+
+      const { emitter } = displayNotification({
+        hash: response.hash,
+      });
+
+      emitter.on("txConfirmed", async() => {
+
+        displayNotification({
+          message: `You staked ${stake.amount.toString()} DOUGH. Your new stake will appear in the list in a really short while...`,
+          type: "success",
+        });       
+
+        const subscription = subject("blockNumber").subscribe({
+          next: async () => {
+            await fetchStakingData();
+            console.log("fetchStakingData", data);               
+
+            subscription.unsubscribe();
+          },
+        });          
+      });
+
+    } catch(error) {
+      console.error(error);
+    }
+  }
+
+  async function unstakeDOUGH(id, lockAmount) {
+    if(!sharesTimeLock) return;
+
+    try {
+      let response = await sharesTimeLock.withdraw(id);
+
+      const { emitter } = displayNotification({
+        hash: response.hash
+      });
+
+      emitter.on("txConfirmed", async() => {
+        displayNotification({
+          message: `You unstaked ${lockAmount.toString()} DOUGH`,
+          type: "success",
+        });
+
+        const subscription = subject("blockNumber").subscribe({
+          next: async () => {
+            console.log("unstakeDOUGH -> blockNumber");
+
+            await fetchStakingData();
+            console.log("fetchStakingData", data);   
+
+            subscription.unsubscribe();
+          },
+        });      
+      });
+
+    } catch(error) {
+      if(error.message.includes('lock not expired')) {
```

Comment:
@Alexintosh i don't really like filtering out errors this way, but the error.message is not a JSON obj we can parse so i did distinguish this way.
Let me know if you have a better/cleaner idea for this.